ubccr/xdmod
# Cleanup / make acl-config more resilient

## Description
- `acl-config`:
  - Added Lock File protection to protect against multiple copies of `acl-config` running simultaneously 
  - Added a signal handler that should ensure the lock file is unlocked when `acl-config` is ctrl-c'd etc. 
  - We no longer need to run `acls-xdmod-management` multiple times within `acl-config`, we now run 
    it once after the backup tables are handled.
  - We've refactored the way that `acl-config` works, previously we wholesale Truncated / re- 
    populated the "managed" tables ( any table that we can completely recreate from available 
    configuration files ) and then re-populated the "non-managed" tables i.e. `user_acls` and 
    `user_acl_group_by_parameters`. This method did not allow us to utilize db transactions as both 
    `TRUNCATE` and DDL ( drop / create / update table ) force an implicit commit on execution. This 
    left us open to having the database in an inconsistent state. Now, instead of truncating / populating 
    the tables directly we instead create / populate a set of `staging` tables from the configuration files 
    first. This leaves the production tables in tact in the event some problem arises. Then, wrapped in 
    a db transaction, we DELETE FROM && INSERT INTO each production table from it's associated 
    staging table. 


## Motivation and Context
We have run into a number of problems lately that we believe to be caused by multiple copies of `acl-config` running at the same time and or interrupted runs of `acl-config`. It's hoped that these changes will resolve these issues. 

## Tests performed
- All automated tests in an OpenXDMoD docker: upgrade / fresh install
- Manual testing: 
  - Inserting an Exception at random times to make sure that existing functionality is not compromised with an incomplete execution. Also, to make sure that we can still utilize the lock file. 
  - Ctrl+C out of an execution at various times to make sure that we don't bork things with an incomplete execution. 

## Types of changes


## Checklist:
